parser/transaction.go

```diff
@@ -325,148 +352,227 @@ func (tx *Transaction) ToCompact(index int) *walletrpc.CompactTx {
 		}
-
-		if !s.ReadCompactSize(&spendCount) {
-			return nil, errors.New("could not read nShieldedSpend")
+		tx.shieldedSpends[i] = newSpend
+	}
+	if !s.ReadCompactSize(&outputCount) {
+		return nil, errors.New("could not read nShieldedOutput")
+	}
+	if outputCount >= (1 << 16) {
+		return nil, errors.New(fmt.Sprintf("outputCount (%d) must be less than 2^16", outputCount))
+	}
+	tx.shieldedOutputs = make([]*output, outputCount)
+	for i := 0; i < outputCount; i++ {
+		newOutput := &output{}
+		s, err = newOutput.ParseFromSlice([]byte(s), tx.version)
+		if err != nil {
+			return nil, errors.Wrap(err, "while parsing shielded Output")
 		}
-
-		if spendCount > 0 {
-			tx.shieldedSpends = make([]*spend, spendCount)
-			for i := 0; i < spendCount; i++ {
-				newSpend := &spend{}
-				s, err = newSpend.ParseFromSlice([]byte(s))
-				if err != nil {
-					return nil, errors.Wrap(err, "while parsing shielded Spend")
-				}
-				tx.shieldedSpends[i] = newSpend
-			}
+		tx.shieldedOutputs[i] = newOutput
+	}
+	if spendCount+outputCount > 0 && !s.ReadInt64(&tx.valueBalanceSapling) {
+		return nil, errors.New("could not read valueBalance")
+	}
+	if spendCount > 0 && !s.Skip(32) {
+		return nil, errors.New("could not skip anchorSapling")
+	}
+	if !s.Skip(192 * spendCount) {
+		return nil, errors.New("could not skip vSpendProofsSapling")
+	}
+	if !s.Skip(64 * spendCount) {
+		return nil, errors.New("could not skip vSpendAuthSigsSapling")
+	}
+	if !s.Skip(192 * outputCount) {
+		return nil, errors.New("could not skip vOutputProofsSapling")
+	}
+	if spendCount+outputCount > 0 && !s.ReadBytes(&tx.bindingSigSapling, 64) {
+		return nil, errors.New("could not read bindingSigSapling")
+	}
+	var actionsCount int
+	if !s.ReadCompactSize(&actionsCount) {
+		return nil, errors.New("could not read nActionsOrchard")
+	}
+	if actionsCount >= (1 << 16) {
+		return nil, errors.New(fmt.Sprintf("actionsCount (%d) must be less than 2^16", actionsCount))
+	}
+	if !s.Skip(820 * actionsCount) {
+		return nil, errors.New("could not skip vActionsOrchard")
+	}
+	if actionsCount > 0 {
+		if !s.Skip(1) {
+			return nil, errors.New("could not skip flagsOrchard")
 		}
-
-		if !s.ReadCompactSize(&outputCount) {
-			return nil, errors.New("could not read nShieldedOutput")
+		if !s.Skip(8) {
+			return nil, errors.New("could not skip valueBalanceOrchard")
 		}
-
-		if outputCount > 0 {
-			tx.shieldedOutputs = make([]*output, outputCount)
-			for i := 0; i < outputCount; i++ {
-				newOutput := &output{}
-				s, err = newOutput.ParseFromSlice([]byte(s))
-				if err != nil {
-					return nil, errors.Wrap(err, "while parsing shielded Output")
-				}
-				tx.shieldedOutputs[i] = newOutput
-			}
+		if !s.Skip(32) {
+			return nil, errors.New("could not skip anchorOrchard")
 		}
-	}
-
-	if tx.version >= 2 {
-		var joinSplitCount int
-		if !s.ReadCompactSize(&joinSplitCount) {
-			return nil, errors.New("could not read nJoinSplit")
+		var proofsCount int
```

Comment:
This is not a count of proofs, it's a size in bytes. Rename it to something like `proofsSize`.